### PR TITLE
Docs network best 2.5

### PR DESCRIPTION
**Pull request by gundalow**
##### SUMMARY
Document best practices for 2.5 Networking, which details:

* Correct way for credentials to be specified
* How to use ``network_cli``
* How to use ``netconf``
* What become mode is, and how to use it

This is best practice. It is assumed readers will have good working Ansible and Ansible Network knowledge by this point

In future PRs the following will be added
* Link to platform specific guides (eAPI, NETCONF, NXAPI, etc)
* A separate guide for Ansible 21.->2.4 will be created

##### ISSUE TYPE
 - Docs Pull Request

### CODE REVIEW

**Comment by IPvSean** [comment]
Thoughts (shouldn't stop the PR, but maybe we can add to another one)

- Can get a list of all available `ansible_network_os` . i.e. ios, nxos, eos, etc where we talk about the `ansible_network_os`
- This might sound dumb... but maybe we can make a picture and put it up there?  I can try drafting up something with you
- Before we dive into group_vars, etc maybe we could have a folder structure for the demo, like literally just do the `tree` command and show that output so they can see our layout?
- Are we using a ansible.cfg file for this?  inventory is not a normal name so I imagine you set this?  Maybe we can take your literal example and put it on our github.com/network-automation github repo !!
- One or more network devices that are compatible with Ansible.  Maybe we link to this-> https://access.redhat.com/articles/3185021
- At the bottom can we link to the network-automation github repo? github.com/network-automation
- The only thing "confusing" to me at all, is the "Collect data" part.  Is that the name of the example playbook or is this where you talk about gathering facts and explaining that?  I feel like you should call the playbook like backup.yml or something and then we put it on github.  I am just confused at the section title Collect data... its just a bit confusing if thats the name of the playbook, a sub-section or what.... maybe its just the way its rendering in markdown versus the docs page

**Comment by gundalow** [comment]
> Can get a list of all available `ansible_network_os` . i.e. ios, nxos, eos, etc where we talk about the `ansible_network_os`

Good point, I think I'll add this to intro_network

> This might sound dumb... but maybe we can make a picture and put it up there?  I can try drafting up something with you

Picture of what?

> Before we dive into group_vars, etc maybe we could have a folder structure for the demo, like literally just do the `tree` command and show that output so they can see our layout?

`tree` is good idea - DONE

> Are we using a ansible.cfg file for this?  inventory is not a normal name so I imagine you set this? 

No ansible.cfg, I'm just doing
`ansible-playbook -i inventory fetch-facts.yml` where `-i` means "use this file for inventory"

> Maybe we can take your literal example and put it on our github.com/network-automation github repo !!

Maybe, will think more on this

>One or more network devices that are compatible with Ansible.  Maybe we link to this-> https://access.redhat.com/articles/3185021

Not seem that page before, though that only lists what has "commercial supported", `dellos*` (for example) supports network_cli, though isn't covered by the "Network Addon"

> At the bottom can we link to the network-automation github repo? github.com/network-automation

Sure, do we have a few lines summary you could provide to explain what is is?

> The only thing "confusing" to me at all, is the "Collect data" part.  Is that the name of the example playbook or is this where you talk about gathering facts and explaining that?  I feel like you should call the playbook like backup.yml or something and then we put it on github.  I am just confused at the section title Collect data... its just a bit confusing if thats the name of the playbook, a sub-section or what.... maybe its just the way its rendering in markdown versus the docs page

Renamed to `facts-demo.yml` (and show in `tree`)

File: docs/docsite/rst/network_best_practices_2.5.rst [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+       - name: Display some facts
+         debug:
+           msg: "The hostname is {{ ansible_net_hostname }} and the OS is {{ ansible_net_version }}"
+
+       - name: Facts from a specific host
+         debug:
+           var: hostvars['vyos01.example.net']
+
+       - name: Write facts to disk using a template
+copy:
```
Comment (Qalthos) [inline docs/docsite/rst/network_best_practices_2.5.rst]: Looks like some indentation got lost?

Comment (Qalthos) [reply docs/docsite/rst/network_best_practices_2.5.rst]: Ah, ansibot saw it, too

**Comment by IPvSean** [comment]
>Picture of what?

Was thinking of super simple diagram since we are not talking about anything but backing up configurations here... but something just to break the wall of text up.picture

>Sure, do we have a few lines summary you could provide to explain what is is?

The Network-Automation GitHub project is a open community for all things network automation.  Here there will be various examples, demonstrations and Github repos dedicated exclusively for automation networking platforms.  To join simply email ansible-network@redhat.com and we will get you added right away!

>Not seem that page before, though that only lists what has "commercial supported", dellos* (for example) supports network_cli, though isn't covered by the "Network Addon"

Yeah I don't know if we have a great list of "everything yet" so maybe just word it like, List of all Red Hat supported networking modules until we get the community supported list as well (or link to the module index page for networking)

**Comment by IPvSean** [comment]
branch base was moved/deprecated, adding new Pull Request

**Comment by gundalow** [comment]
* Removed platform specific notes, this will be documented on pages linked from `intro_network.rst`
* Corrected link to `intro_network`

@IPvSean I'll put a diagram on `intro_network.rst`

**Comment by ismc** [comment]
It seems confusing to use ansible_network_os at times and the group (when grouped by os) at others, especially when you are setting ansible_network_os in group_vars.

File: docs/docsite/rst/network_best_practices_2.5.rst [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+
+.. warning:: Never store passwords in plain text
+
+The "Vault" feature of Ansible allows keeping sensitive data such as passwords or keys in encrypted files, rather than as plain text in your playbooks or roles. These vault files can then be distributed or placed in source control. See :doc:`playbooks_vault` for more information.
+
+:ansible_connection:
+
+  The ansible-connection setting tells Ansible how it should connect to a remote device. When working with Ansible Networking, setting this to ``ansible_connection: network_cli`` informs Ansible that the remote node is a network device with a limited execution environment. Without this setting, Ansible would attempt to use ssh to connect to the remote and execute the Python script on the network device, which would fail because Python generally isn't available on network devices.
+:ansible_network_os:
+  Informs Ansible which Network platform this hosts corresponds to. This is required when using ``network_cli``
```
Comment (privateip) [inline docs/docsite/rst/network_best_practices_2.5.rst]: required for both `network_cli` and `netconf`

File: docs/docsite/rst/network_best_practices_2.5.rst [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+     hosts: switches
+     gather_facts: no
+
+     tasks:
+       ###
+       # Collect data
+       #
+       - name: Gather facts (eos)
+         eos_facts:
+         when: "'eos' in group_names"
```
Comment (privateip) [inline docs/docsite/rst/network_best_practices_2.5.rst]: `when: ansible_network_os == 'eos'`

File: docs/docsite/rst/network_best_practices_2.5.rst [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+
+       - name: Gather facts (ops)
+         ios_facts:
+         when: "'ios' in group_names"
+
+       - name: Gather facts (vyos)
+         vyos_facts:
+         when: "'vyos' in group_names"
```
Comment (privateip) [inline docs/docsite/rst/network_best_practices_2.5.rst]: as above (except vyos)

File: docs/docsite/rst/network_best_practices_2.5.rst [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+       ###
+       # Collect data
+       #
+       - name: Gather facts (eos)
+         eos_facts:
+         when: "'eos' in group_names"
+
+       - name: Gather facts (ops)
+         ios_facts:
+         when: "'ios' in group_names"
```
Comment (privateip) [inline docs/docsite/rst/network_best_practices_2.5.rst]: as above (except ios)

File: docs/docsite/rst/network_best_practices_2.5.rst [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+
+       ###
+       # Get running configuration
+       #
+
+       - name: Backup switch (eos)
+         eos_config:
+           backup: yes
+         register: backup_eos_location
+         when: "'eos' in group_names"
```
Comment (privateip) [inline docs/docsite/rst/network_best_practices_2.5.rst]: `when:` as above

**Review by privateip — CHANGES_REQUESTED** [review]

**Comment by dharmabumstead** [comment]
Doc review in process. Please don’t merge.

File: docs/docsite/rst/network_best_practices_2.5.rst [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+       - name: Facts from a specific host
+         debug:
+           var: hostvars['vyos01.example.net']
+
+       - name: Write facts to disk using a template
+           copy:
+           content: |
+             #jinja2: lstrip_blocks: True
+             EOS device info:
+               {% for host in groups['eos'] %}
```
Comment (gundalow) [inline docs/docsite/rst/network_best_practices_2.5.rst:244]: Is there a way of rewriting this using `ansible_network_os`?

Comment (gundalow) [reply docs/docsite/rst/network_best_practices_2.5.rst:244]: This is apparently a fine use of groups still.

File: docs/docsite/rst/network_best_practices_2.5.rst [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+
+An ``inventory`` file is an INI-like configuration file that defines the mapping of hosts into groups.
+
+In our example, the inventory file defines the groups ``eos``, ``ios``, ``vyos`` and a "group of groups" called ``switches``. Further details about subgroups and inventory files can be found in the :ref:`Ansible inventory Group documentation <subgroups>`.
+
+
+Connection & Credentials (group_vars)
+-------------------------------------
+
+As Ansible is a flexible tool there are a number of ways of specifying connection information & credentials. We believe however that the best was is to use ``group_vars``.
```
Comment (dharmabumstead) [inline docs/docsite/rst/network_best_practices_2.5.rst]: Please revise to: Because Ansible is a flexible tool, there are a number of ways to specify connection information and credentials. We recommend using ``group_vars``.

File: docs/docsite/rst/network_best_practices_2.5.rst [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+                     39336231636137663964343966653162353431333566633762393034646462353062633264303765
+                     6331643066663534383564343537343334633031656538370a333737656236393835383863306466
+                     62633364653238323333633337313163616566383836643030336631333431623631396364663533
+                     3665626431626532630a353564323566316162613432373738333064366130303637616239396438
+                     9853
+
+
+.. FIXME FUTURE Gundalow - Link to network auth & proxy page (to be written)
+
+.. warning:: Never store passwords in plain text
```
Comment (dharmabumstead) [inline docs/docsite/rst/network_best_practices_2.5.rst]: Period

File: docs/docsite/rst/network_best_practices_2.5.rst [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+
+The "Vault" feature of Ansible allows keeping sensitive data such as passwords or keys in encrypted files, rather than as plain text in your playbooks or roles. These vault files can then be distributed or placed in source control. See :doc:`playbooks_vault` for more information.
```
Comment (dharmabumstead) [inline docs/docsite/rst/network_best_practices_2.5.rst]: "allows you to keep" rather than "keeping sensitive data"

File: docs/docsite/rst/network_best_practices_2.5.rst [hunk 11]
```diff
@@ -0,0 +1,372 @@
...
+
+:ansible_connection:
+
+  The ansible-connection setting tells Ansible how it should connect to a remote device. When working with Ansible Networking, setting this to ``ansible_connection: network_cli`` informs Ansible that the remote node is a network device with a limited execution environment. Without this setting, Ansible would attempt to use ssh to connect to the remote and execute the Python script on the network device, which would fail because Python generally isn't available on network devices.
```
Comment (dharmabumstead) [inline docs/docsite/rst/network_best_practices_2.5.rst]: Suggest revising to: Ansible uses the ``ansible-connection`` setting to determine how to connect to a remote device. When working with Ansible Networking, set this to netweork_cli so Ansible treats the remote node as a network device with a limited execution environment.

File: docs/docsite/rst/network_best_practices_2.5.rst [hunk 12]
```diff
@@ -0,0 +1,372 @@
...
+Ansible ``hostvars`` allows you to access variables from a named host. Without this we would return the details for the current host, rather than the named host.
+
+For more information, see :ref:`magic_variables_and_hostvars`.
+
+Get running configuration
+-------------------------
+
+The :ref:`eos_config <eos_config>` and :ref:`vyos_config <vyos_config>` modules have a ``backup:`` option that when set will cause the module to create a full backup of the current ``running-config`` from the remote device before any changes are made. The backup file is written to the ``backup`` folder in the playbook root directory. If the directory does not exist, it is created.
+
+To demonstrate how we can move the backup file to a different location we ``register`` the result and use the ``backup_path`` return value as source location to move the file into ``/tmp/backups/`` directory which we have created.
```
Comment (dharmabumstead) [inline docs/docsite/rst/network_best_practices_2.5.rst]: Hard to parse. If I'm understanding it correctly, maybe try: "To demonstrate how we can move the backup file to a different location, we ``register`` the result and move the file to the path stored in ``backup_path``."

File: docs/docsite/rst/network_best_practices_2.5.rst [hunk 13]
```diff
@@ -0,0 +1,372 @@
...
+
+Note that when using variables from tasks in this way we use double quotes (``"``) and double curly-brackets (``{{...}}`` to tell Ansible that this is a variable.
+
+Troubleshooting
+===============
+
+If you receive an connection error please double check the inventory and Playbook for typos or missing lines, if the issue still occurs follow the debug steps in :doc:`network_debug_troubleshooting`.
```
Comment (dharmabumstead) [inline docs/docsite/rst/network_best_practices_2.5.rst]: Period instead of comma.

**Review by dharmabumstead — CHANGES_REQUESTED** [review]
A few suggestions. Thanks @gundalow!

**Review by Qalthos — APPROVED** [review]

**Comment by gundalow** [comment]
@dharmabumstead Updated based on your feedback, thanks

Ready for final review

**Review by dharmabumstead — APPROVED** [review]
Looks good. Thanks @gundalow!

**Review by privateip — APPROVED** [review]

**Comment by gundalow** [comment]
rebuild_merge
